Cache labels.Labels to Identify Series in Remote Write

csmarchbanks opened the pull request:
This is an experiment to have the remote write component use `labels.Labels` instead of `[]prompb.Label` when passing samples around/identifying series. A `label.Label` uses less space so steady state memory usage should be less, with the tradeoff of transforming `label.Labels` to `[]prompb.Label` for each remote request.

Results after removing some extra sources of allocations look very good. This is for a server with three remote-write targets, and ~8k samples per second ingestion.

### Benchmarks
```
benchmark                     old ns/op     new ns/op     delta
BenchmarkSampleDelivery-8     998914        1007751       +0.88%

benchmark                     old allocs     new allocs     delta
BenchmarkSampleDelivery-8     7117           6117           -14.05%

benchmark                     old bytes     new bytes     delta
BenchmarkSampleDelivery-8     688391        640400        -6.97%
```

### Heap profiles
Heap profiles show a 50% reduction in memory used by StoreSeries as expected.

#### v2.11.1:
![v2 11 1](https://user-images.githubusercontent.com/3280472/62817409-4fca7b80-baf3-11e9-9a7b-41374b458568.png)

#### rw-use-labels branch
![rw-use-labels](https://user-images.githubusercontent.com/3280472/62817410-50631200-baf3-11e9-97b4-a83e4697a0a8.png)

### Other metrics
CPU usage is very similar between the servers. However, there is a reduction in allocations:
![ReducedAllocations](https://user-images.githubusercontent.com/3280472/62817450-3413a500-baf4-11e9-8e1e-48495836d7b7.png)

brian-brazil commented on storage/remote/queue_manager.go:
```diff
@@ -552,13 +555,19 @@ func (s *shards) start(n int) {
 
 	s.queues = newQueues
 
+	s.tsPool = sync.Pool{
```
Does this really help? Changing the number of shards isn't that often.

csmarchbanks replied:
Benchmarks improved some when I first added it. The pool is used for every http request, not just the number of shards, though it could probably be done by keeping a single `[]prompb.TimeSeries` per shard instead of using a pool, similar to the compression buffer.

brian-brazil commented on storage/remote/codec.go:
```diff
@@ -372,9 +372,16 @@ func labelsToLabelsProto(labels labels.Labels) []prompb.Label {
 	result := make([]prompb.Label, 0, len(labels))
```
You can be smarter here I think, I presume this is where the allocations are.

csmarchbanks replied:
:+1: Many were here, and some still in buildWriteRequest from making `[]prompb.Sample`. Latest update helps both of these.

brian-brazil commented on storage/remote/queue_manager.go:
```diff
@@ -748,16 +752,17 @@ func (s *shards) sendSamplesWithBackoff(ctx context.Context, samples []prompb.Ti
-		if ts.Samples[0].Timestamp > highest {
-			highest = ts.Samples[0].Timestamp
+	req := s.writeRequests.get(len(samples))
+	defer s.writeRequests.put(req)
```
defer causes an alloc. I'd avoid sync.Pool entirely, it has interesting interactions with GC and in this case we know exactly how much memory we need.

csmarchbanks replied:
:+1: Removed the pool.

brian-brazil commented:
That's looking okay. What do the benchmarks say now?

csmarchbanks commented:
I removed the merge conflict, and updated the benchmarks, slightly fewer allocations than last time, CPU jumps around between runs but is about the same before/after.

I don't expect the graphs to change too much, but can repost them in a few hours if interested.

csmarchbanks commented:
@brian-brazil would you mind taking another look when you have a moment?

brian-brazil commented:
:+1:

cstyan commented on storage/remote/codec.go:
```diff
@@ -192,14 +192,15 @@ func StreamChunkedReadResponses(
 		series := ss.At()
 		iter := series.Iterator()
-		lbls := MergeLabels(labelsToLabelsProto(series.Labels()), sortedExternalLabels)
+		lbls = MergeLabels(labelsToLabelsProto(series.Labels(), lbls), sortedExternalLabels)
```
looks like remote read only uses `labelsToLabelsProto` in tests so it should be okay

cstyan commented on storage/remote/queue_manager.go:
```diff
@@ -545,11 +540,17 @@ func (t *QueueManager) newShards() *shards {
 	return s
 }
 
+type sample struct {
```
Hmmm, would be nice if we could use a different name. To me sample implies just ts and value, but I can't think of anything better right now.

csmarchbanks replied:
I could call it something like `queueSample` to differentiate it a bit, but I don't know if that gains anything.

cstyan commented on storage/remote/codec.go:
```diff
@@ -522,13 +523,25 @@ func labelProtosToLabels(labelPairs []prompb.Label) labels.Labels {
 	return result
 }
+
+func internLabels(lbls labels.Labels) {
```
release is there too

csmarchbanks replied:
:+1:
